### ISSUE TITLE
iOS Undefined symbols architecture with latest update

### ISSUE THREAD

**Issue by marcanw** [distriqt/ANE-GooglePlayServices#28]
Compiling from Animate (AIR 50.2.3.1) on Windows for iOS get this error since I updated GooglePlayServices:

![image](https://github.com/distriqt/ANE-GooglePlayServices/assets/21979056/bc8d8f5c-24d3-4f9b-b7ec-f275463f478d)

**Comment by marchbold** [comment]
Looks like we need to get Harman to update the stubs packaged with the AIR SDK again. I'll look into it. 

In the meantime either use macOS to package your app or revert to the previous version.

**Comment by marchbold** [comment]
Which extensions are you using that is causing this error btw?

**Comment by marcanw** [comment]
> Looks like we need to get Harman to update the stubs packaged with the AIR SDK again. I'll look into it.
> 
> In the meantime either use macOS to package your app or revert to the previous version.

I reverted back. 
Can I update the other ane?
![image](https://github.com/distriqt/ANE-GooglePlayServices/assets/21979056/6a7dcbd4-8004-414d-92ad-1012b14099ee)

**Comment by marcanw** [comment]
> Which extensions are you using that is causing this error btw?
I would say Inapp-Billing, according to the errors in the screen of the original post.

**Comment by marchbold** [comment]
I believe that is actually the core google firebase dependency in the error, so likely one of the extensions that depend on that. 

The others you listed should be okay to update for ios, as long as you aren't using the FCM variant of push notifications.

**Comment by marcanw** [comment]
> I believe that is actually the core google firebase dependency in the error, so likely one of the extensions that depend on that.
> 
> The others you listed should be okay to update for ios, as long as you aren't using the FCM variant of push notifications.
I use the FCM version. So I better avoid to update till there is a solution. 
I ,just updated AndroidSupport. Is it fine?

**Comment by marchbold** [comment]
I would use the older versions of the androidx libs as well if you are trying to keep the same versions on android

**Comment by marcanw** [comment]
Hello @marchbold 
Do you know when I will have a solution for this?

**Comment by marchbold** [comment]
Hi, I have been discussing potential solutions for this with the Harman team. The main issue is that the Apple opensource compiler for iOS apps that AIR uses hasn't been updated by Apple in a while and has fallen behind the features available in Xcode. So AIR is facing an issue with packaging iOS apps on Windows that use recent features in iOS or libraries that have been packaged with the latest build tools.

As there seems to be developers who don't want to or can't use macOS, I was wondering what you all would think of a hosted packaging solution for developers on Windows where a host provided by Harman would package your application for iOS and provide the ipa as an alternative. There would obviously be a cost involved (potentially part of the AIR license), but this would allow you to continue developing iOS applications on Windows. 

Would love to hear your thoughts on this idea?

(Note this doesn't affect developers using macOS for iOS development)

**Comment by ajwfrost** [comment]
Well, there's some more information about the cause of all this, from the LLVM project:
https://github.com/llvm/llvm-project/issues/56034

I'd just looked at the symbols exported by a framework with/without this flag, and there are zero differences other than those symbols being there! But of course, within the executable code, these symbols will be called in order to call each of the selectors...

So, it may feasibly be possible to build the "latest" Apple linker on Windows, and merge in the changes that LLVM have made, and then have this support these newly built libraries. But building the Apple linker on Windows was proving to be very complex already..

The alternative option -> providing a service whereby ADT would upload the object files to an online service, backed by a macOS computer that just does the linking -> might actually be simpler! but perhaps not quicker. And scalability may be an issue (and service availability, given how often our macs reboot themselves!)

**Comment by Ender22** [comment]
I'm a bit confused because I am packaging on mac with intelliJ but the latest com.google.firebase.core.ane ANE version is also giving me the same errors. 

Does the latest require an update to XCode or something? I'm using a bit of an older xcode...

**Comment by sjabberwocky** [comment]
@Ender22 I also found that the latest update of Firebase does not compile and previous crush app on launch on iOS 12.5.5.

**Comment by marchbold** [comment]
@Ender22 Yes you need to update the compiler on macOS, easiest done by updating Xcode. 

@sjabberwocky I'll follow up on the crash in the Firebase repo.

**Comment by Ender22** [comment]
Is it possible to update the compiler without updating xcode? I had a lot of trouble packaging on mac before, and right now it's in a 'if it isn't broke don't fix it' state.

I've downloaded xcode13 but can't install it without updating MacOS first. Could I like, just replace the path to the compiler to the new xcode13 one maybe?
 
If it doesn't work like that and would be a big hassle then I'll just bite the bullet and update the OS and xcode, but thought I'd ask first.

**Comment by ajwfrost** [comment]
We've tried things like that before (there's a Developer/Toolchains folder in Xcode, iirc) and it worked to a certain extent but there may then be other issues you find with the compiler features, expected header files, and all sorts... it can get messy. So it could be worth trying, but there may be challenges... Apple tend to expect everyone updates all the time to the latest of everything...

**Comment by Ender22** [comment]
Understood, @marchbold what is the lowest xcode version that works with this ane? Apparently, my Mac is too old to install Ventura, so the best I can do is Monterey which only supports XCode 13.4 it seems

**Comment by Ender22** [comment]
Have tested it on XCode 13.4.1 and it seems to package fine, yay. Thanks guys

**Comment by Ender22** [comment]
Wait crap no. It doesn't compile on XCode 13.4.1, I tested the wrong ane version. So I'll need to buy a new Mac for this. Can I go back to the question of what other ANE's are expected to fail if I use the older com.google.firebase.core? Maybe it affects some functionality I don't actually use?

**Comment by marchbold** [comment]
@Ender22 The hidden costs of Apple development hey... 

I would suggest if you are trying to maintain an older version that you use APM to install versions prior to this latest update. Don't just install a lower version of `com.google.firebase.core` but pull a version of all the extensions at a specific release. If you were using apm your lock file would have contained the details and you could just revert your version control to that version? Otherwise you can try just downloading the version from the github repositories

**Comment by Ender22** [comment]
Yes, I'm using APM. 

Hmm okay, looking at my lock file has left me a bit confused. I use the following extensions:

- Application
- Application Rater
- Camera Roll Extended
- CameraUI
- Dialog
- Flurry
- Media Player
- Native WebView
- Network Info
- Notifications
- PDF Reader
- Push Notifications
- Scanner
- Share

Looks like com.google.firebase.core is used in:
com.distriqt.playservices.CloudMessaging
com.google.android.datatransport
and those are part of google play services right?... and I think from my list of extensions, I don't really understand from my list of extensions which is using google play services, we don't try to do anything directly with that.

Additionally, since google play services aren't used on iOS, I guess nothing will stop working? 

Sorry, I'm a bit lost. Do you know offhand which one of my extensions is causing the firebase.core requirement, so I could roll back just that specific extension?

**Comment by marchbold** [comment]
Okay, so looks as though you could probably remove this extension for the ios build. It is required for Firebase services which is likely getting included as you are using Push Notifications which uses FCM on Android. We are working on getting platform dependant builds working with apm but it's only partially implemented so far. 

Currently you could try either manually removing it, or setting up an alternative apm project config for ios and installing the APNS variant of the push notifications extension.

**Comment by marcanw** [comment]
@marchbold 
Due to this issue, I'm stuck with older ANEs.

I have two questions:
1. Are the latest versions of the ANEs required for Android 33?
2. As I understand it, the problem arises when building iOS on Windows, but this concerns an ANE that is only required for Android (referring to your last post). Perhaps I'm overlooking something. Could you clarify?

Thank you for your response.

**Comment by marchbold** [comment]
@marcanw 
1. I believe the previous version of the majority of our extensions supported API 33. This latest update is to support the latest Google services (Firebase/Analytics/FCM etc) which included several bug fixes from Google.
2. `com.google.firebase.core` is required by iOS apps supporting Firebase services so not an Android only extension.

**Comment by idzdigital** [comment]
Can someone please post here once the issue is solved and ANEs are ready for development on Windows.
Thanks

**Comment by Michoko92** [comment]
Michael pointed me to this thread, and the discussion about possibly dropping the iOS support on Windows. Honestly, this AIR feature is probably the most precious and unique to me, compared to other SDKs. Every time I check new SDKs or engines, the first thing I look for is how to build for iOS target. Until now, all the alternatives I saw can only build on an Apple machine. As an experienced Windows user, being able to build all platforms on my dev machine is a comfort only AIR offers, and it is invaluable.

I'm sure that, knowing Apple, maintaining iOS support must be a very challenging task. But honestly, this is something that could definitely be one of the strongest selling points for AIR once marketing is done more actively. Please tell me there is still hope.

PS: I suppose that having a remote building service would be OKish, and it would be better than nothing. But I think that if AIR drops Windows iOS building in a future version, I'll stop updating as long as I can.

**Comment by ajwfrost** [comment]
Hi

Thanks @Michoko92 for the input. The support for iOS on Windows is a tricky one, but we think we may have a route forwards for it. To link in another thread: the below is our tracking one for the 'undefined symbols' problem when building on Windows.
https://github.com/airsdk/Adobe-Runtime-Support/issues/2299

We're investigating still - particularly the updates that may be coming along with Xcode 15 - but I would say that there is now some hope! We do now have some updated ld64 code from Apple which we hope supports the msgsend selector stubs, so will be looking into this now.

thanks

**Comment by MatseFR** [comment]
Hi,
Michael pointed me to this thread as well.
I do have an old mac mini, I updated it and it's now running Monterey and XCode 14.2

I'm not a mac user at all but after hours of hairpulling I finally managed to run ADT with the latest AIR SDK... only to get the same "Undefined symbols for architecture arm64" error, which is quite disappointing to be honest.

Do I need a mac that can run XCode 15 ?

Here is the command I used, do I miss something in there ?

adt -package -target ipa-app-store -provisioning-profile appstore.mobileprovision -storetype pkcs12 -keystore ios_dist.p12 -storepass xxxxx app.ipa application.xml -C icons/ios/ . -C bin . -extdir ane -platformsdk /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.2.sdk/

**Comment by ajwfrost** [comment]
Hi

Definitely don't update to Xcode 15..  the fact you're getting this on a mac is odd though .. possibly you could remove the "platformsdk" argument in case that's confusing things? 

Is your mac mini an Intel one? And are the missing symbols all ones that have "objc_msgsend" in the name?

thanks

**Comment by Michoko92** [comment]
Thank you @ajwfrost , this is excellent news! Keep up the amazing work, you and your team are really appreciated. 🙏

**Comment by MatseFR** [comment]
@ajwfrost Thanks for replying, I get the same error when removing the "platformsdk" argument

My mac is an intel one yes, however there is no occurence of "objc_msgsend" in the error log (attached to this post), a lot of FirebaseXXX occurences though (I do use Firebase for push notifications using distriqt's ANE)
[error_log.txt](https://github.com/distriqt/ANE-GooglePlayServices/files/12665797/error_log.txt)

**Comment by marchbold** [comment]
@MatseFR This looks like a separate issue, open one in the Firebase repository and we'll help you through it. Definitely remove the platform sdk option, and make sure you have v14 of Xcode installed (do not install 15 yet!)

**Comment by MatseFR** [comment]
@marchbold I couldn't install Xcode 15 even if I wanted to : my Mac is "too old" (thanks Apple for programmed obsolescence !)

I don't use the Firebase ANE (at least not directly) : I use FCM for push notifications, shouldn't I open an issue in the Push Notifications repo instead ?

**Comment by marchbold** [comment]
Yeah PushNotifications extension would be appropriate then.

**Comment by idanasher** [comment]
I want to buy a new mac in order to be able to compile...
I'm debating between two mac(s)  and I would love to get a recommendation\suggestion if it matters to the development process and also would like to know which operating system, and XCode version I should have on this new mac I want to buy.

These are the option:

(1) Mac mini M2 Pro Chip
(2) Mac Studio M2 Max

**Comment by MatseFR** [comment]
@idanasher Right now you need XCode 14.3, which requires at least macOS 13

You can install previous versions of XCode from this link (requires an apple dev account) https://developer.apple.com/download/all/

If you end up with several versions of XCode installed you can choose which one should be used by command line tools such as ADT by launching XCode and going to 
Settings > Locations > Command line tools

Also if you are a windows user like me and want to buy a mac just to package IPAs you might want to consider other possibilities first : if you have a friend/colleague/whatever with a mac that is recent enough and that person is ok to let you use it you can just use remote desktop software such as https://rustdesk.com/ (which is extremely easy to use) to package your IPA. That's how I did it a few days ago using my brother's mac

**Comment by marchbold** [comment]
@idanasher Unless you need the extra power of the studio a mini is perfectly fine for AIR development. We actually use a mac mini as a build server for a range of projects, including AIR and Unity.

**Comment by marcanw** [comment]
Is there any hope that Harman will finally find a solution for Windows, or is it hopeless?

**Comment by ajwfrost** [comment]
Well we got the "latest" code for the linker from Apple .. but that's from Xcode 13 it seems, and doesn't handle these objc_msgSend stubs. So we're looking to see whether we can generate an object file that would contain the definitions... which is an interesting challenge! And once we've done that, we'd need to then get this all building on Windows.

I wouldn't say hopeless, but it's technically quite challenging.. I'll try to provide an update at the end of next week to let you know whether we've managed to get it working on the mac platform, and if so then it's "just" a case of getting the Apple linker to build on Windows. It would really help though if Apple released the Xcode 14 linker source code :-(

**Comment by marcanw** [comment]
Thx @ajwfrost for your update. I hope you guys will find a solution.

**Comment by ajwfrost** [comment]
@marcanw @Michoko92  can you please try downloading the folder from the below link, this should replace the `lib\aot\bin\ld64` folder in your AIR SDK. This contains a new build of the linker, using LLVM's rather than Apple's source - it's worked in our internal testing but it would be good to know if this also runs and builds/links for your apps..

https://transfer.harman.com/link/BxsZukLF5pKkU6XKAzcVO6

thanks

**Comment by Michoko92** [comment]
Hi @ajwfrost , so I first downloaded the latest version of the AIR SDK (50.2.3.6), without the new linker, and when I tried an iOS build, I got this error message:

ld: unknown option: -platform_version

Then I downloaded the files above, and replaced the linker, and now everything seems to work fine. Good job!

Just wanted to tell you that some people may encounter the same -platform_version error with the vanilla 50.2.3.6 version, though?

Also I want to precise I was not encountering initially the same issues as @marcanw , as I am not using the ANEs that may trigger this error. But at least regular iOS building works fine for me . Cheers!

**Comment by ajwfrost** [comment]
Thanks - yes it's a bit in flux at the moment, we may need to do a quick update to 50.2.3.6 to update the linker command for Windows prior to pushing out a version that actually uses the new linker.. sounds like there are still some issues to sort out with that :-(

**Comment by marcanw** [comment]
I started to test a build that was compiling flawlessly with 50.2.3.5 before upgrading to last ANEs that was creating problems.

As [Michoko92](https://github.com/Michoko92), I had the same issue with the vanilla 50.2.3.6:
![image](https://github.com/distriqt/ANE-GooglePlayServices/assets/21979056/2f3f554a-19f1-4d74-996a-a134be9bc5ef)

After replacing ld64, I get the following error:
![image](https://github.com/distriqt/ANE-GooglePlayServices/assets/21979056/ced1ed9d-a7cd-412e-9959-c191e8782719)

I don't know if I migrate to the last ANEs it would solve the problem. Maybe @marchbold knows more about it.

**Comment by marchbold** [comment]
@marcanw That is a separate issue, I've got a fix for it I'm rolling out shortly.

**Comment by marcanw** [comment]
@marchbold Ok, so if I update the Firebase ANE, it will solve the issue, right?

**Comment by mrfrasier** [comment]
> @marcanw @Michoko92 can you please try downloading the folder from the below link, this should replace the `lib\aot\bin\ld64` folder in your AIR SDK. This contains a new build of the linker, using LLVM's rather than Apple's source - it's worked in our internal testing but it would be good to know if this also runs and builds/links for your apps..
> 
> https://transfer.harman.com/link/BxsZukLF5pKkU6XKAzcVO6
> 
> thanks

Any chance for an updated link? I'm also facing the same issue and popped in to see if there was an update, but seems the file transfer has expired

**Comment by ajwfrost** [comment]
@mrfrasier  there's an updated version - this link will work hopefully for the next month...
https://transfer.harman.com/link/Brh2YO2NAELPL99TZbgimo
We need to push out a couple of releases to get this properly out into the SDK, had hoped to get that done before month-end but it will be a little delayed..

**Comment by mrfrasier** [comment]
Thanks @ajwfrost, that resolved the platform issue, but I'm still unable to build. Looks like there's a bunch of issues with duplicate symbols and Bolt?

<img width="798" alt="image" src="https://github.com/distriqt/ANE-GooglePlayServices/assets/4335972/5f54c54d-cfce-41d3-8eb8-3c92cb99e4d1">

<img width="938" alt="image" src="https://github.com/distriqt/ANE-GooglePlayServices/assets/4335972/58c9514c-9264-4563-b6f4-108f24d33c85">

**Comment by ajwfrost** [comment]
The "updated" one here was the LLVM linker with updates to switch errors (caused by duplicate symbols) into warnings, the Apple ld64 linker seems to ignore these.

But actually, those all should just be outputs that warn about duplicates, but the final linkage/output should still work? Are there any lines that start with 'error'?

**Comment by mrfrasier** [comment]
@ajwfrost sadly I can't scroll far enough up to see the first lines, but all the others are warnings. The final lines are as such:

<img width="934" alt="image" src="https://github.com/distriqt/ANE-GooglePlayServices/assets/4335972/bae152de-b387-4b1f-b832-0df3b847667e">

**Comment by marchbold** [comment]
> @marchbold Ok, so if I update the Firebase ANE, it will solve the issue, right?

Yes, update `com.google.firebase.core`

**Comment by ajwfrost** [comment]
@mrfrasier you can hopefully enable the ADT debug log - or simpler, just launch the AIR SDK Manager and go to the "troubleshooting" tab. Then if you re-run it, the log should appear in that pane, or in the c:\users\marcus\adt.log file..

**Comment by mrfrasier** [comment]
@ajwfrost thanks for that. Going step by step I've managed to isolate it to the Firebase ANE causing the trouble (somehow I can run the PushNotifications-FCM ANE without it), so this may be a @marchbold issue specific to the ANE.

In any case, here's the actual error that kills things:

`ld64: error: undefined symbol: _APMUserDataFieldEmailAddress
>>> referenced by E:\\tmp\\0870d453-7e84-49fd-b942-93791733fa84\FirebaseAnalytics.framework\FirebaseAnalytics(FIRAnalytics.o):(symbol +[FIRAnalytics initiateOnDeviceConversionMeasurementWithEmailAddress:]+0x5c)
>>> referenced by E:\\tmp\\0870d453-7e84-49fd-b942-93791733fa84\FirebaseAnalytics.framework\FirebaseAnalytics(FIRAnalytics.o):(symbol +[FIRAnalytics initiateOnDeviceConversionMeasurementWithEmailAddress:]+0x58)
`

After that is a bunch of warnings as you saw before.

**Comment by marchbold** [comment]
@mrfrasier the com.distriqt.Firebase ANE is not a requirement for FCM in the com.distriqt.PushNotifications extension. We have kept these extensions independent so you don't need to purchase both. 

Can you log this issue in the Firebase repo and we'll look into it.

**Comment by marcanw** [comment]
> @marcanw @Michoko92 can you please try downloading the folder from the below link, this should replace the `lib\aot\bin\ld64` folder in your AIR SDK. This contains a new build of the linker, using LLVM's rather than Apple's source - it's worked in our internal testing but it would be good to know if this also runs and builds/links for your apps..
> 
> https://transfer.harman.com/link/BxsZukLF5pKkU6XKAzcVO6
> 
> thanks

Hi @ajwfrost
There is a new 50.2.3.7 release. Do we have to download the folder again for this new version?

**Comment by ajwfrost** [comment]
@marcanw yes, sorry, we did the 50.2.3.7 version just with Android updates, so if you're using it for iOS you'd need to update the linker again -> new download link is https://transfer.harman.com/link/Brh2YO2NAELPL99TZbgimo

We'll be pushing out 50.2.4.1 shortly that has these plus the other updates needed on iOS for the latest Xcode/SDKs..

**Comment by idanasher** [comment]
Trying to compile my app on mac for the first time using animate.

I am getting this Launch-Screen related issue:

![WhatsApp Image 2023-11-07 at 09 22 34](https://github.com/distriqt/ANE-GooglePlayServices/assets/8189469/b416c991-7880-418f-b4db-74bedce4deb1)

**Comment by marchbold** [comment]
@idanasher Not sure that is related to this issue. Try opening a new discussion in the AIR forum: https://github.com/airsdk/Adobe-Runtime-Support/discussions/new/choose

**Comment by marcanw** [comment]
@marchbold & @ajwfrost 
Just to let you know I updated all my ANE's and the last AIR (50.2.3.8) and all my problems are gone.
Thanks for the great job :-)

**Comment by idanasher** [comment]
> @marchbold & @ajwfrost Just to let you know I updated all my ANE's and the last AIR (50.2.3.8) and all my problems are gone. Thanks for the great job :-)

on a mac \ pc ?

**Comment by marcanw** [comment]
PC

**Comment by idanasher** [comment]
Hearing the news, I am back on my pc trying to compile my ios app with 50.2.3.8.

first, Do I still need to include the 'Framework' folder on my app root like I have always did ?

second: i have successfully compiled the app but when trying to upload the app via mac transporter I get those rejects:


> ITMS-90338: Non-public API usage - The app contains one or more corrupted binaries. Rebuild the app and resubmit.. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed. For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/
> 
> ITMS-90482: Invalid Executable - The executable '[rummy_world.app/Frameworks/FBAEMKit.framework/FBAEMKit](http://rummy_world.app/Frameworks/FBAEMKit.framework/FBAEMKit)' contains bitcode.
> 
> ITMS-90482: Invalid Executable - The executable '[rummy_world.app/Frameworks/FBSDKCoreKit.framework/FBSDKCoreKit](http://rummy_world.app/Frameworks/FBSDKCoreKit.framework/FBSDKCoreKit)' contains bitcode.
> 
> ITMS-90482: Invalid Executable - The executable '[rummy_world.app/Frameworks/FBSDKCoreKit_Basics.framework/FBSDKCoreKit_Basics](http://rummy_world.app/Frameworks/FBSDKCoreKit_Basics.framework/FBSDKCoreKit_Basics)' contains bitcode.
> 
> ITMS-90482: Invalid Executable - The executable '[rummy_world.app/Frameworks/FBSDKGamingServicesKit.framework/FBSDKGamingServicesKit](http://rummy_world.app/Frameworks/FBSDKGamingServicesKit.framework/FBSDKGamingServicesKit)' contains bitcode.
> 
> ITMS-90482: Invalid Executable - The executable '[rummy_world.app/Frameworks/FBSDKLoginKit.framework/FBSDKLoginKit](http://rummy_world.app/Frameworks/FBSDKLoginKit.framework/FBSDKLoginKit)' contains bitcode.
> 
> ITMS-90482: Invalid Executable - The executable '[rummy_world.app/Frameworks/FBSDKShareKit.framework/FBSDKShareKit](http://rummy_world.app/Frameworks/FBSDKShareKit.framework/FBSDKShareKit)' contains bitcode.
>

**Comment by marchbold** [comment]
@idanasher Have you updated the Facebook ANE? Looks like you may be using an older version that still contains bitcode.

**Comment by idanasher** [comment]
@ajwfrost  Regarding

> Just to check, when they talk about "the private Apple APIs listed above", did you see any APIs listed?

No, i did not get any more information on this.
and thank you for your response Andrew


@marchbold 
i have used Facebook ane version 16.01 and i am now getting the latest version16.0.101.
Do you recommend any more specific ane's latest versions to update ?  
Thanks a lot Michael

[ this is what i use now:
Core.VERSION== 7.2.0
Starling.VERSION==2.5.1
NetworkInfo VERSION = 4.0.12
Application.VERSION==6.11.0
Facebook.instance.version==16.0.1
Branch.VERSION ===5.0.067
Adverts.VERSION== 14.0.3
InAppBilling.version ==14.4.1

]

**Comment by marchbold** [comment]
I suggest updating them all. Particularly InAppBilling and the Core ANE. Also [16.0.103](https://github.com/distriqt/ANE-FacebookAPI/releases/tag/v16.0.103) is the latest release of the Facebook ANE not 16.0.101

**Comment by idanasher** [comment]
i have replaced all ane's to the latest version and still getting the same 'ITMS-90482: Invalid Executable   Contain bitcode ' issues

**Comment by marchbold** [comment]
@idanasher Did you update the frameworks?

**Comment by idanasher** [comment]
Nope
I have not updated it in a long time

Where do I het it from ?
Thanks a lot

**Comment by yvant** [comment]
> second: i have successfully compiled the app (and then re-signed it on my mac with the resign script but when trying to upload the app via mac transporter I get those rejects:
> 
> > I also build my ipa on Windows and didn't hear about that resigning script. Where can I find it?

**Comment by marchbold** [comment]
@idanasher it's in the repo alongside the extensions. You must always update all associated assets and frameworks when updating an extension. 

@yvant You shouldn't need the resign script anymore, AIR has corrected the issue that required this

**Comment by yvant** [comment]
Ok thanks for the info. I had a "Non-public API usage" error and I thought it could be related. I've seen a thread with a similar bug, I'll just follow that other one.

**Comment by ajwfrost** [comment]
@yvant yes for an issue with binaries being rejected by Apple for "Non-public API usage", please refer to:
https://github.com/airsdk/Adobe-Runtime-Support/issues/2911

Essentially, we had switched to using the LLVM mach-o linker in order to support the new object file formats that are used by Xcode 14, but these are producing output that - while it runs fine - are for some reason unacceptable to Apple. So for the short term we'll need to switch back to using the Apple linker, which is fine on macOS (as it exists!) but on Windows the latest linker Apple have published is based on Xcode 13 so we'd end up with the same issues as found in the original problem report above.

Actually - breaking news - Apple have just updated their open source drops! https://github.com/apple-oss-distributions/ld64/tree/ld64-907

So we'll get working on that ....

**Comment by idanasher** [comment]
@marchbold 

> it's in the repo alongside the extensions.

Can't find it.

**Comment by marcanw** [comment]
HI @ajwfrost 
Any idea about the timing on this Apple rejection?
Did you think about asking a grace period to Apple?
Does it necessarily occur on each IPA build with AIR 50.2.4.1 on Windows?

**Comment by marchbold** [comment]
@marcanw Is that related to this issue?

**Comment by ajwfrost** [comment]
Hmm.. define "this issue" ..! Originally this was about the objc_msgSend symbols not being worked out by the Adobe-build Apple linker. We've kind of resolved that by switching to the LLVM macho linker, but that then has this issue where Apple are rejecting the binaries that are then created.

The binaries appear to work.. I don't know whether anyone has tried to challenge Apple on this, given that their message says that the app uses some non-public APIs as listed below, but then fails to list anything. It's very possible their response would just be to suggest using their own linker rather than LLVM... so our expectation is that the best approach is to use the Apple linker itself.

There's then the problem that when we use the Apple linker from Xcode 15 (on mac) it doesn't work if you link against the 'stub' files we generate. So we're now using the linker from Xcode 14 (ld-classic) when compiling on a mac, and the LLVM macho linker when compiling on Windows. We end up with binaries that both run fine, but only the mac-built ones are accepted into the App Store.

I'm hoping that we can build the Apple linker for Windows (which would be the Xcode 14 one), and then I think we may also need to do away with the 'stubs' folder and have people use the iPhoneOS SDK directly, so that the correct set-up of symbols between different frameworks on different iOS versions works properly, and so that the macOS builds can then use the latest/Xcode 15 linker..

**Comment by marchbold** [comment]
Ah thanks for the update, wasn't aware there was a rejection occurring with the update, hence my confusion :)

**Comment by Mintonist** [comment]
Hm... I use AIR 50.2.4.1 on Windows and my apps have been approved some days ago.

**Comment by mrfrasier** [comment]
> Hm... I use AIR 50.2.4.1 on Windows and my apps have been approved some days ago.

Are you sure your submission was built with 50.2.4.1? Numerous folks (myself included) have gotten rejections with that exact version. If so, it may end up coming down to what APIs (libraries) your app actually accesses, which is interesting.

**Comment by Mintonist** [comment]
Sorry I use AIR 50.2.3.8 but I check the same version of ld64 binaries (\lib\aot\bin\ld64) with AIR 50.2.4.1

And I see this issue was commented while AIR 50.2.3.8 was the latest. 

I don't want use AIR 50.2.4.1 because of IOS version restriction (11->12)

**Comment by mrfrasier** [comment]
@ajwfrost For the Windows LLVM issue, since there is no ETA, is there a simple method to export the project to Mac for building? We've been hoping to push a fix to a build last updated in October :/

**Comment by marcanw** [comment]
@ajwfrost or if you have a better view on the timing because I'm really stuck...
Thx

**Comment by ajwfrost** [comment]
@mrfrasier we can work on something for that, it should be something we can work around...
@marcanw do you have any mac that you could use for linking? anything that supports Xcode 14 would be fine..

I'm thinking that if you're using Windows as the primary development machine, we could have a set-up where you also have a mac that runs a bit like a Jenkins node, i.e. a client machine that is triggered to build the software by a "master" server. But with a very specific/limited case, to make it a little simpler..

thanks

**Comment by mrfrasier** [comment]
@ajwfrost sounds good. I don't have an active mac anymore (I mean, I guess I could dust off my old macbook), but I use a VirtualBox instance to submit apps to the store. Hopefully that should suffice

**Comment by marcanw** [comment]
@ajwfrost I tried building on Windows with the latest AIR version 50.2.4.3, and I was able to publish successfully. I'm not sure if it's due to the latest AIR version, or if I was never affected by the "ITMS-90482: Invalid Executable" issue from the start.

**Comment by ajwfrost** [comment]
Thanks @marcanw  - when you say "publish", do you mean it's been accepted for the App Store? We appear to have two separate issues with the Windows linking:
1. when using symbols that are present across different frameworks dynamically, depending on the iOS version, it typically leads to a crash-on-startup when running on an older device
2. when publishing for the App Store, Apple seem to mark the binary as containing private APIs and block it as an invalid/corrupt executable

@mrfrasier I'm quite impressed you managed to get macOS running within a VirtualBox instance! does that have a reasonably recent version on it, e.g. with Xcode 14 or later?

thanks

**Comment by mrfrasier** [comment]
@ajwfrost I took a look and it seems my VM is running Catalina and xcode 11.3.1 -- though the app store says the latest is 15.2 but doesn't give the option to update.

It says there's a system update available, so I'll try that

**Comment by marcanw** [comment]
@ajwfrost 
I guess I wasn't concerned by 2. I was able to publish on App Store two times

**Comment by Mintonist** [comment]
> when using symbols that are present across different frameworks dynamically, depending on the iOS version, it typically leads to a crash-on-startup when running on an older device

@ajwfrost where you discuss this problem and what status for this now?
It is very important for my projects! Thanks!

**Comment by mrfrasier** [comment]
> > when using symbols that are present across different frameworks dynamically, depending on the iOS version, it typically leads to a crash-on-startup when running on an older device
> 
> @ajwfrost where you discuss this problem and what status for this now? It is very important for my projects! Thanks!

Pushing this back to the forefront. I haven't been able to publish to iOS < 16 since last year. I still have a lot of users asking me about this regularly. @ajwfrost do we have an update on that front? We got an updated SDK that *should* work, but then the issues with not being able to submit to the app store arose. What is the current method for getting our apps out there?
